### PR TITLE
Slack alert throttle

### LINKED ISSUE #14: Slack alerts option multiple alerts
It just occurred to me that the `--slack-alerts` option will result in a Slack message every hour while that job is running.
It would be useful to store a record of JWDs for which notifications have been sent to prevent duplicate notifications.

### PR DESCRIPTION
**Pull request by neoformit**
Resolves #14 

- Throttle slack alerts, defaults to 1 per week per JWD
- Keep a history of slack notifications in temp file (defaults to /tmp/walle-notifications.txt)

Not sure if these should live in the Ansible repo? I could just gitignore them?
- Add some tests
- Add requirements.txt

### CODE REVIEW

**Comment by mira-miracoli** [comment]
> Not sure if these should live in the Ansible repo? I could just gitignore them?

Sorry I don't really see why, maybe you can explain more about that? (or maybe it is obvious but I did not see it) also we need to make sure there is no user data involved.
Is it not possible that this lives on the node?

**Comment by neoformit** [comment]
>maybe you can explain more about that?

It's only that the test.py and requirements.txt files are for development and not something that needs to be `ansible-galaxy` installed by everyone. I don't mind if they live in here if you're happy!

>make sure there is no user data involved.

If you mean the `NotificationHistory` class, it only stores date strings and JWD path strings in the history file.

**Comment by mira-miracoli** [comment]
ah okay, sorry I misunderstood and thought that you wanted to store the slack history in the github repo.
yes the requirements and tests of course can live here  and also it is nice to have the requirements if people use the virtualenv method! :)

**Comment by mira-miracoli** [comment]
Sorry I have another question:
Was there a specific reason you used the JWD for the notification history and not the `galaxy_id` (job id)?

File: files/walle.py [hunk 1]
```diff
@@ -56,6 +63,67 @@ def convert_arg_to_seconds(hours: str) -> float:
...
+        return [line[1] for line in self._read_records()]
+
+    def _read_records(self) -> List[List[str]]:
+        with open(self.record_file, "r") as f:
+            records = [
+                line.strip().split("\t") for line in f.readlines() if line.strip()
+            ]
+        return self._validate(records)
+
+    def _validate(self, records: List[List[str]]) -> List[List[str]]:
```
Comment (mira-miracoli) [inline files/walle.py:85]: maybe we could add better type hints here, would make it more readable.

**Comment by neoformit** [comment]
> Sorry I have another question: Was there a specific reason you used the JWD for the notification history and not the `galaxy_id` (job id)?

Nope, it was just the most obvious identifier to me at the time. Job ID would be fine too!

File: files/test.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import unittest
```
[the remark this answers is not in the document]
Comment (neoformit) [reply files/test.py:1]: Sure, and it would be great to expand the test coverage, especially since it's so awkward to test manually!

**Comment by neoformit** [comment]
@mira-miracoli happy for me to merge this?

File: files/walle.py [hunk 3]
```diff
@@ -63,6 +64,20 @@ def convert_arg_to_seconds(hours: str) -> float:
     return float(hours) * 60 * 60
 
 
+@dataclass
+class Record:
```
Comment (mira-miracoli) [inline files/walle.py:68]: cool, thanks a lot 😊

**Review by mira-miracoli — APPROVED** [review]
Yes! Sorry for being annoying and nit-picky

File: files/test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mira-miracoli) [reply files/test.py:1]: absolutely!

**Pull request merged**